README.md
# ReTo - Render auxiliary Tools
(c) 2016. GMRV / URJC

www.gmrv.es
gmrv@gmrv.es

## Introduction

ReTo provides the following tools:

* Camera
* Shader manager
* Picking functionality
* Object Loader (triangle/quad mode)
* Spline navigation
* TextureManager

## Dependencies

As required dependencies ReTo needs:
* OpenGL
* GLEW
* Eigen3

The following dependencies are optional:
* GLUT for some examples
* ZeroEQ/Lexis for publishing camera
* Boost's Unit Test Framework for unit testing.
* FreeImage for Texture2D using file images.

## Building

ReTo has been succesfully built and used on Ubuntu 14.04/16.04, Mac OSX
Yosemite and Windows 7/8 (Visual Studio 2013 Win64). The following steps
should be enough to build it:

```bash
git clone --recursive https://gitlab.gmrv.es/nsviz/ReTo.git
mkdir ReTo/build && cd ReTo/build
cmake .. -DCMAKE_BUILD_TYPE=Release
make
```

## Testing

After building ReTo, you can run some examples in bin directory. Some of the
examples need to find the shaders through an environment variable pointing to
the examples directory (available on both the source and installation trees).
For example, from the building directory, created as explained before, you can run:
~~~~~~~~~~~~~
RETO_SHADERS_PATH=../examples/ bin/ReToDemoCubes
RETO_SHADERS_PATH=../examples/ bin/ReToTextureReader ../examples/gmrv_grande.png
~~~~~~~~~~~~~

## Documentation

You can access the online API documentation generated from the source (using
[Doxygen](http://doxygen.org/)) in https://gmrvvis.github.io/doc/ReTo/


Changelog.md
# ReTo - Render auxiliary Tools

## git master

## 0.3.1

	*[!43] Added rubberband, lasso and clipping planes capabilities.

## 0.3.0
	* [!39] Added setters for several camera attributes.
	* [!36] Release 0.3. Updated README.md. Changed CMake/common to submodule. Fixed examples shaders path.

## 0.2.0
	* [!35] Fixed Camera code with no ZeroEQ. RETO_USE_ZEROEQ renamed to RETO_USE_LEXIS
	* [!34] Improving tests
	* [!33] Fixed OSX clang warnings
	* [!31] Fixed Python script
	* [!30] Check glut for tests.
	* [!29] Fixed shaders generation
	* [!28] Fixed reto shaderprogram xiv
	* [!27] Remove glut header from TextureManager.cpp (made windows build fail)
	* [!25] Fixed compilation for Ubuntu 16.04.
	* [!24] textureManager
	* [!21] Added TextureManager
	* [!18] Picking and program improvement. Added Glut Demo. Added Picking Demo. Added Transform Feedback varying in ShaderProgram. Added autocatching to ShaderProgram (only catch active uniforms and attributes.
	* [!19] Shaders module
	* [!17] Fixed reto_generate_shaders warnings
	* [!15] Obj loader
	* [!16] Fixed encoding errors in old python version (<3.0)
	* [!13] Added Spline
	* [!9] Add warning in shaders locations data, Update PickingSystem. Add shaders generation files (Python)
	* [!11] Added reto_generate_shaders(.cmake and .py).
	* [!10] Moved defines to cmake options and fixed examples. Minor styling and cleanups.
	* [!7] Cleanups. Fixes for nvidia detection and for ZeroEQ and Lexis
	* [!6] Better way of resolving nvidia/mesa conflict.
	* [!4] Camera doc
	* [!2] Simple pickable2


examples/cube.h
/*
 * Copyright (c) 2014-2016 GMRV/URJC.
 *
 * Authors: Cristian Rodriguez
 *
 * This file is part of ReTo <https://gitlab.gmrv.es/nsviz/ReTo>
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License version 3.0 as published
 * by the Free Software Foundation.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this library; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 */
#ifndef __RETO__CUBE__
#define __RETO__CUBE__

const int cubeNVertex = 24;
const int cubeNTriangleIndex = 12;

const unsigned int cubeTriangleIndex[] =
{
	//Face z = 1
	0,1,2,			1,3,2,
	//Face z = -1
	4,6,5,			5,6,7,
	//Face x = 1
	8,10,9,			9,10,11,
	//Face x = -1
	12,13,14,		13,15,14,
	//Face y = 1
	16,17,18,		17,19,18,
	//Face y = -1
	20,22,21,		21,22,23,
};

const float cubeVertexPos[] =
{
	//Face z = 1
	-1.0f,	-1.0f,	 1.0f, //0
	 1.0f,	-1.0f,	 1.0f, //1
	-1.0f,	 1.0f,	 1.0f, //2
	 1.0f,	 1.0f,	 1.0f, //3

	//Face z = -1
	-1.0f,	-1.0f,	-1.0f, //4
	 1.0f,	-1.0f,	-1.0f, //5
	-1.0f,	 1.0f,	-1.0f, //6
	 1.0f,	 1.0f,	-1.0f, //7

	//Face x = 1
	1.0f,	-1.0f,	-1.0f, //8
	1.0f,	-1.0f,	 1.0f, //9
	1.0f,	 1.0f,	-1.0f, //10
	1.0f,	 1.0f,	 1.0f, //11

	//Face x = -1
	-1.0f,	-1.0f,	-1.0f, //12
	-1.0f,	-1.0f,	 1.0f, //13
	-1.0f,	 1.0f,	-1.0f, //14
	-1.0f,	 1.0f,	 1.0f, //15

	//Face y = 1
	-1.0f,	 1.0f,	-1.0f, //16
	-1.0f,	 1.0f,	 1.0f, //17
	 1.0f,	 1.0f,	-1.0f, //18
	 1.0f,	 1.0f,	 1.0f, //19

	//Face y = -1
	-1.0f,	-1.0f,	-1.0f, //20
	-1.0f,	-1.0f,	 1.0f, //21
	 1.0f,	-1.0f,	-1.0f, //22
	 1.0f,	-1.0f,	 1.0f  //23
 };

const float cubeVertexNormal[] =
{
	//Face z = 1
	0.0f,	0.0f,	 1.0f,
	0.0f,	0.0f,	 1.0f,
	0.0f,	0.0f,	 1.0f,
	0.0f,	0.0f,	 1.0f,

	//Face z = -1
	0.0f,	0.0f,	-1.0f,
	0.0f,	0.0f,	-1.0f,
	0.0f,	0.0f,	-1.0f,
	0.0f,	0.0f,	-1.0f,

	//Face x = 1
	1.0f,	0.0f,	 0.0f,
	1.0f,	0.0f,	 0.0f,
	1.0f,	0.0f,	 0.0f,
	1.0f,	0.0f,	 0.0f,

	//Face x = -1
	-1.0f,	0.0f,	 0.0f,
	-1.0f,	0.0f,	 0.0f,
	-1.0f,	0.0f,	 0.0f,
	-1.0f,	0.0f,	 0.0f,

	//Face y = 1
	0.0f,	1.0f,	0.0f,
	0.0f,	1.0f,	0.0f,
	0.0f,	1.0f,	0.0f,
	0.0f,	1.0f,	0.0f,

	//Face y = -1
	0.0f,	-1.0f,	0.0f,
	0.0f,	-1.0f,	0.0f,
	0.0f,	-1.0f,	0.0f,
	0.0f,	-1.0f,	0.0f
 };

const float cubeVertexColor[] =
{
	//Face z = 1
	0.0f,	0.0f,	 1.0f,
	0.0f,	0.0f,	 1.0f,
	0.0f,	0.0f,	 1.0f,
	0.0f,	0.0f,	 1.0f,

	//Face z = -1
	0.6f,	0.6f,	 1.0f,
	0.6f,	0.6f,	 1.0f,
	0.6f,	0.6f,	 1.0f,
	0.6f,	0.6f,	 1.0f,

	//Face x = 1
	1.0f,	0.0f,	 0.0f,
	1.0f,	0.0f,	 0.0f,
	1.0f,	0.0f,	 0.0f,
	1.0f,	0.0f,	 0.0f,

	//Face x = -1
	1.0f,	0.6f,	 0.6f,
	1.0f,	0.6f,	 0.6f,
	1.0f,	0.6f,	 0.6f,
	1.0f,	0.6f,	 0.6f,

	//Face y = 1
	0.0f,	1.0f,	0.0f,
	0.0f,	1.0f,	0.0f,
	0.0f,	1.0f,	0.0f,
	0.0f,	1.0f,	0.0f,

	//Face y = -1
	0.6f,	1.0f,	0.6f,
	0.6f,	1.0f,	0.6f,
	0.6f,	1.0f,	0.6f,
	0.6f,	1.0f,	0.6f
 };

const float cubeVertexTexCoord[] =
{
	//Face z = 1
	 0.0f, 0.0f,
	 1.0f, 0.0f,
	 0.0f, 1.0f,
	 1.0f, 1.0f,

	//Face z = -1
	0.0f, 1.0f,
	1.0f, 1.0f,
	0.0f, 0.0f,
	1.0f, 0.0f,

	//Face x = 1
	0.0f,	1.0f,
	1.0f,	1.0f,
	0.0f,	0.0f,
	1.0f,	0.0f,

	//Face x = -1
	0.0f,	0.0f,
	1.0f,	0.0f,
	0.0f,	1.0f,
	1.0f,	1.0f,

	//Face y = 1
	0.0f, 1.0f,
	0.0f, 0.0f,
	1.0f, 1.0f,
	1.0f, 0.0f,

	//Face y = -1
	0.0f, 0.0f,
	0.0f, 1.0f,
	1.0f, 0.0f,
	1.0f, 1.0f,
};

#endif /* __RETO__CUBE__ */
